url43a0865bb0_kevinjdolan_darwin_tgz-pJ8-tui_DriverJ8/src/darwInvest/data/gathering/dataSource/ReutersNews.java

```diff
@@ -0,0 +1,188 @@
+			return true;
+		}
+		else return false;
+	}
+
+	@Override
+	public void gatherData() {
+		//no of stories to read in, for time's sake.
+		int nostories = 15;
+		File target = new File("StockData/" + source);
+		target.mkdir();
+		File listoftickers = new File("RawData/" + super.getName() + "/tickers.txt");
+		Scanner s = null;
+		try {
+			s = new Scanner(new BufferedReader(new FileReader(listoftickers)));
+			s.useDelimiter("[\n\r]+");
+			while(s.hasNext()) {
+				String currentstock = s.next();
+				Ticker ticker = new Ticker(currentstock);
+				String companyname = ticker.lookupCompanyname();
+				String reuters = "http://www.reuters.com/search?blob=%22" + webize(companyname) +  "%22";
+				readIn(ticker, reuters, nostories);
+				addTicker(ticker);
+			}
+		} catch (FileNotFoundException e) {
+			e.printStackTrace();
+		} catch (MalformedURLException e) {
+			e.printStackTrace();
+		} catch (IOException e) {
+			e.printStackTrace();
+		} finally {
+			if (s != null) {
+				s.close();
+			}
+		}
+	}
+
+	private void readIn(Ticker ticker, String reuters, int nostories) {
+		Boolean reallydone = false;
+		int storycounter = 0;
+		int pagenumber = 1;
+		while (!reallydone && storycounter < nostories) {
+			try {
+				String nexturl = reuters.concat("&pn=" + pagenumber);
+				URL reutersurl = new URL(nexturl);
+				BufferedReader reutersreader = new BufferedReader(new InputStreamReader(reutersurl.openStream()));
+				System.out.println(ticker.getSymbol() + ":" + nexturl);
+				Scanner s3 = new Scanner(reutersreader);
+				s3.useDelimiter("[\n\r]+");
+				Boolean done = false;
+				while(s3.hasNext() && !done && storycounter < nostories) {
+					String next = s3.next();
+					if(next.contains("<div class=\"secondaryContent\">")) {
+						done = true;
+					}
+					if(next.contains("<div class=\"searchResult\">")) {
+						String extractor = s3.next();
+						String extractor2 = extractor.substring(extractor.indexOf("href"));
+						String title = extractor2.substring(extractor2.indexOf('>')+1,extractor2.indexOf('<'));
+						URL exploreurl = new URL(extractor2.substring(extractor2.indexOf('\"')+1,extractor2.indexOf('>')-1) + "?sp=true");
+						s3.next();
+						String dateextractor = s3.next().substring(2);
+						String da = dateextractor.substring(dateextractor.indexOf('>')+1,dateextractor.indexOf('<'));
+						Date date = dconv(da);
+						//Test timestamp
+						if (inRange(startdate,enddate,date)) {
+							String content = URLExplorer(exploreurl);
+							ticker.addNews(new NewsEvent(title, content, date.getTime()));
+						}
+						else if (date.before(startdate)) {
+							reallydone = true;
+							break;
+						}
+						storycounter++;
+					}
+				}
+				pagenumber++;
+				if (s3 != null) {
+					s3.close();
+				}
+				if (reutersreader != null) {
+					reutersreader.close();
+				}
+			} catch (Exception e) {
+				e.printStackTrace();
+			}
+		}
+	}
+
+	private String URLExplorer(URL exploreurl) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 178-185 into a separate method.